### PR TITLE
Deprecate default service IP CIDR

### LINKED ISSUE #81315: DefaultServiceIPCIDR should not exist
API server defaults the service CIDR.  We should instead just fail if it is not specified.  Defaulting is a wild guess and very likely to be wrong.

/sig network
/kind bug

### PR DESCRIPTION
**Pull request by darshanime**
**What type of PR is this?**
/kind bug
**What this PR does / why we need it**:
Remove default service CIDR

**Which issue(s) this PR fixes**:
Fixes #81315 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
ACTION REQUIRED:
Deprecate the default service IP CIDR. The previous default was `10.0.0.0/24` which will be removed in 6 months/2 releases. Cluster admins must specify their own desired value, by using `--service-cluster-ip-range` on kube-apiserver.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs

```

/sig network

### CODE REVIEW

**Comment by liggitt** [comment]
note that this requires all existing deployments that rely on the default to modify their invocation to replace it with the current default value. that's a change that needs to be pre-announced and rolled out per https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-flag-or-cli

File: pkg/master/services.go [hunk 1]
```diff
@@ -21,28 +21,23 @@ import (
...
 
-// DefaultServiceIPRange takes a the serviceIPRange flag and returns the defaulted service ip range (if  needed),
-// api server service IP, and an error
-func DefaultServiceIPRange(passedServiceClusterIPRange net.IPNet) (net.IPNet, net.IP, error) {
-	serviceClusterIPRange := passedServiceClusterIPRange
-	if passedServiceClusterIPRange.IP == nil {
-		klog.Infof("Network range for service cluster IPs is unspecified. Defaulting to %v.", kubeoptions.DefaultServiceIPCIDR)
-		serviceClusterIPRange = kubeoptions.DefaultServiceIPCIDR
+// GenericAPIServerServiceIP takes a the serviceIPRange flag and returns the api server service IP, and an error
+func GenericAPIServerServiceIP(serviceClusterIPRange net.IPNet) (net.IP, error) {
```
Comment (liggitt) [inline pkg/master/services.go]: this isn't generic, it only applies to kube-apiserver, possibly just rename to `ServiceIPRange`?

File: pkg/master/services.go [hunk 2]
```diff
@@ -21,28 +21,23 @@ import (
...
+	if serviceClusterIPRange.IP == nil {
+		return net.IP{}, fmt.Errorf("Network range for service cluster IPs is unspecified: %q", serviceClusterIPRange)
```
Comment (liggitt) [inline pkg/master/services.go]: to help people avoid messing up their existing clusters, we should probably include the old default in the error message and indicate what value an upgraded cluster should use

File: pkg/master/services.go [hunk 3]
```diff
@@ -21,28 +21,23 @@ import (
...
-	kubeoptions "k8s.io/kubernetes/pkg/kubeapiserver/options"
 	"k8s.io/kubernetes/pkg/registry/core/service/ipallocator"
 )
 
-// DefaultServiceIPRange takes a the serviceIPRange flag and returns the defaulted service ip range (if  needed),
-// api server service IP, and an error
-func DefaultServiceIPRange(passedServiceClusterIPRange net.IPNet) (net.IPNet, net.IP, error) {
-	serviceClusterIPRange := passedServiceClusterIPRange
-	if passedServiceClusterIPRange.IP == nil {
-		klog.Infof("Network range for service cluster IPs is unspecified. Defaulting to %v.", kubeoptions.DefaultServiceIPCIDR)
```
Comment (liggitt) [inline pkg/master/services.go]: for the first N releases while we wait out the deprecation period, we should elevate this to a warning, and indicate in the message that the invocation should be changed to specify this explicitly

**Comment by darshanime** [comment]
@liggitt if I understand the deprecation policy correctly, we should emit a warning and continue setting the default cidr for 1 release/6 months?

**Comment by liggitt** [comment]
Emit a warning if the flag is unset and default it to the current default, add the deprecation notice to the release notes until two releases (6 months) have passed

File: cmd/kube-apiserver/app/options/options.go [hunk 4]
```diff
@@ -114,7 +114,6 @@ func NewServerRunOptions() *ServerRunOptions {
 		},
 		ServiceNodePortRange: kubeoptions.DefaultServiceNodePortRange,
 	}
-	s.ServiceClusterIPRange = kubeoptions.DefaultServiceIPCIDR
```
Comment (darshanime) [inline cmd/kube-apiserver/app/options/options.go]: Removing assignment from here so it goes thru `ServiceIPRange` with a warning.

**Comment by thockin** [comment]
This needs a more complete release note, otherwise the change LGTM.

Please update the relnote and unhold

/lgtm
/approve

File: pkg/master/services.go [hunk 5]
```diff
@@ -25,12 +25,12 @@ import (
 	"k8s.io/kubernetes/pkg/registry/core/service/ipallocator"
 )
 
-// DefaultServiceIPRange takes a the serviceIPRange flag and returns the defaulted service ip range (if  needed),
+// ServiceIPRange takes a the serviceIPRange flag and returns the defaulted service ip range (if  needed),
```
Comment (beejeebus) [inline pkg/master/services.go]: I noticed a couple of typos in the docs, s/takes a the/takes the/, and a double space in 'if  needed'.

Beyond the typos, perhaps the docs could be:

```
// ServiceIPRange checks if the serviceIPRange flag is nil, setting service ip
// range to the default value in kubeoptions.DefaultServiceIPCIDR in that case.
// Returns service ip range, api server service IP, and an error
```

Comment (darshanime) [reply pkg/master/services.go]: made the change. 
/retest

**Comment by jeanluclariviere** [comment]
@darshanime Do you know how this change will effect existing kubeadm initialized clusters? Specifically, when upgrading from one minor version to another? Typically, changes made to the kube-apiserver need to be persisted to the kubeadm-config configmap - if however we initialized our cluster **without** DefaultServiceIPCIDR, would this effect the running services?

In other words, after the default value has been removed, what CIDR will existing services use once we upgrade to 1.20? Do we need to explicitly set the **DefaultServiceIPCIDR** in the kubeadm-config configmap?

Thanks!

**Pull request merged**